### PR TITLE
Transition to v2 shims [Databricks]

### LINKED ISSUE #4474: Transition away from v1 shims
**Is your feature request related to a problem? Please describe.**
It's confusing for a developer to write shims especially if they are used to developing using the v1 way. 

**Describe the solution you'd like**
Remove the v1 usage

### PR DESCRIPTION
**Pull request by razajafri**
This PR in a nutshell 

* Moved all the SparkShims to a standard package `com.nvidia.spark.rapids.shims.v2` so we don't have to dynamically look them up at runtime
* Combined duplicate classes to reduce code redundancy
* Removed the need to call the `ShimLoader` when accessing the Shims

fixes #4474

### CODE REVIEW

**Comment by tgravescs** [comment]
so one thing to be aware of here is that any of our partners that have shim layers are going to have to transition as well. which means more work on their side so lets try to get this done how we want in one release.

File: sql-plugin/src/main/303/scala/com/nvidia/spark/rapids/shims/spark303/Spark303Shims.scala [hunk 1]
```diff
@@ -1,42 +0,0 @@
-/*
- * Copyright (c) 2021-2022, NVIDIA CORPORATION.
```
Comment (tgravescs) [inline sql-plugin/src/main/303/scala/com/nvidia/spark/rapids/shims/spark303/Spark303Shims.scala:2]: why was this one removed and not moved?  m

File: sql-plugin/src/main/304/scala/com/nvidia/spark/rapids/v2/SparkShims.scala [hunk 2]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2021-2022, NVIDIA CORPORATION.
+ * Copyright (c) 2020-2022, NVIDIA CORPORATION.
```
Comment (tgravescs) [inline sql-plugin/src/main/304/scala/com/nvidia/spark/rapids/v2/SparkShims.scala:2]: should be 2021

File: sql-plugin/src/main/304/scala/org/apache/spark/sql/rapids/shims/v2/RapidsShuffleInternalManager.scala [hunk 3]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2021, NVIDIA CORPORATION.
+ * Copyright (c) 2019-2022, NVIDIA CORPORATION.
```
Comment (tgravescs) [inline sql-plugin/src/main/304/scala/org/apache/spark/sql/rapids/shims/v2/RapidsShuffleInternalManager.scala:2]: should be 2021

File: sql-plugin/src/main/311until320-all/scala/org/apache/spark/sql/rapids/shims/v2/RapidsShuffleInternalManager.scala [hunk 4]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2021, NVIDIA CORPORATION.
+ * Copyright (c) 2019-2022, NVIDIA CORPORATION.
```
Comment (tgravescs) [inline sql-plugin/src/main/311until320-all/scala/org/apache/spark/sql/rapids/shims/v2/RapidsShuffleInternalManager.scala:2]: 2021

File: sql-plugin/src/main/312-nondb/scala/com/nvidia/spark/rapids/shims/spark312/SparkShimServiceProvider.scala [hunk 5]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2021, NVIDIA CORPORATION.
+ * Copyright (c) 2020-2022, NVIDIA CORPORATION.
```
Comment (tgravescs) [inline sql-plugin/src/main/312-nondb/scala/com/nvidia/spark/rapids/shims/spark312/SparkShimServiceProvider.scala:2]: 2021

File: sql-plugin/src/main/31xdb/scala/com/nvidia/spark/rapids/shims/v2/AQEUtils.scala [hunk 6]
```diff
@@ -18,13 +18,11 @@ package com.nvidia.spark.rapids.shims.v2
...
 import org.apache.spark.sql.catalyst.expressions.Attribute
 import org.apache.spark.sql.execution.adaptive.{QueryStageExec, ShuffleQueryStageExec}
-import org.apache.spark.sql.execution.exchange.ReusedExchangeExec
 
 /** Utility methods for manipulating Catalyst classes involved in Adaptive Query Execution */
 object AQEUtils {
   /** Return a new QueryStageExec reuse instance with updated output attributes */
   def newReuseInstance(sqse: ShuffleQueryStageExec, newOutput: Seq[Attribute]): QueryStageExec = {
-    val reusedExchange = ReusedExchangeExec(newOutput, sqse.shuffle)
-    ShuffleQueryStageExec(sqse.id, reusedExchange, sqse.originalPlan)
```
Comment (tgravescs) [inline sql-plugin/src/main/31xdb/scala/com/nvidia/spark/rapids/shims/v2/AQEUtils.scala:28]: this change seems out of place and not related to shim work?

File: sql-plugin/src/main/31xdb/scala/com/nvidia/spark/rapids/shims/v2/Spark31XdbShimsBase.scala [hunk 7]
```diff
@@ -63,7 +63,7 @@ trait Spark31XdbShimsBase extends SparkShims {
   override def newBroadcastQueryStageExec(
       old: BroadcastQueryStageExec,
       newPlan: SparkPlan): BroadcastQueryStageExec =
-    BroadcastQueryStageExec(old.id, newPlan, old.originalPlan)
+    old.newReuseInstance(old.id, newPlan.output).asInstanceOf[BroadcastQueryStageExec]
```
Comment (tgravescs) [inline sql-plugin/src/main/31xdb/scala/com/nvidia/spark/rapids/shims/v2/Spark31XdbShimsBase.scala:66]: same here, this doesn't seem related to shim work?

File: sql-plugin/src/main/320/scala/com/nvidia/spark/rapids/shims/v2/RapidsShuffleManager.scala [hunk 8]
```diff
@@ -0,0 +1,26 @@
+/*
+ * Copyright (c) 2021-2022, NVIDIA CORPORATION.
```
Comment (tgravescs) [inline sql-plugin/src/main/320/scala/com/nvidia/spark/rapids/shims/v2/RapidsShuffleManager.scala:2]: no 2021 if new file

File: sql-plugin/src/main/322/scala/com/nvidia/spark/rapids/shims/v2/RapidsShuffleManager.scala [hunk 9]
```diff
@@ -0,0 +1,26 @@
+/*
+ * Copyright (c) 2021-2022, NVIDIA CORPORATION.
```
Comment (tgravescs) [inline sql-plugin/src/main/322/scala/com/nvidia/spark/rapids/shims/v2/RapidsShuffleManager.scala:2]: no 2021

File: sql-plugin/src/main/330/scala/com/nvidia/spark/rapids/v2/RapidsShuffleManager.scala [hunk 10]
```diff
@@ -0,0 +1,26 @@
+/*
+ * Copyright (c) 2021-2022, NVIDIA CORPORATION.
```
Comment (tgravescs) [inline sql-plugin/src/main/330/scala/com/nvidia/spark/rapids/v2/RapidsShuffleManager.scala:2]: do we need all these individual shuffle manager files, are some the same that we could put in a 3XXuntil3YY?

Comment (tgravescs) [reply sql-plugin/src/main/330/scala/com/nvidia/spark/rapids/v2/RapidsShuffleManager.scala:2]: also no 2021

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/ShimLoader.scala [hunk 11]
```diff
@@ -53,13 +53,13 @@ import org.apache.spark.util.MutableURLClassLoader
 
     E.g., Spark 3.2.0 Shim will use
 
-    jar:file:/home/spark/rapids-4-spark_2.12-22.04.0.jar!/spark3xx-common/
-    jar:file:/home/spark/rapids-4-spark_2.12-22.04.0.jar!/spark320/
```
Comment (tgravescs) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/ShimLoader.scala:57]: revert

File: pom.xml [hunk 12]
```diff
@@ -115,6 +115,7 @@
                                     <sources>
                                         <source>${project.basedir}/src/main/301+-nondb/scala</source>
                                         <source>${project.basedir}/src/main/301/scala</source>
+                                        <source>${project.basedir}/src/main/301until304/scala</source>
```
[the remark this answers is not in the document]
Comment (razajafri) [reply pom.xml:118]: It doesn't apply to 3.0.4. The signature of `getParquetFilters` changed in 3.0.4 which is why there is a separate folder for 304

File: sql-plugin/src/main/301db/scala/com/nvidia/spark/rapids/shims/spark301db/SparkShimServiceProvider.scala [hunk 13]
```diff
@@ -16,7 +16,7 @@
 
 package com.nvidia.spark.rapids.shims.spark301db
```
[the remark this answers is not in the document]
Comment (razajafri) [reply sql-plugin/src/main/301db/scala/com/nvidia/spark/rapids/shims/spark301db/SparkShimServiceProvider.scala:17]: I can get rid of the `v2` from the package name. 

Yes, you are right the service provider is the only class to remain in the version specific package. 

The README.md from shims module got moved to [docs/dev/shims.md](https://github.com/NVIDIA/spark-rapids/blob/branch-22.04/docs/dev/shims.md)

File: sql-plugin/src/main/303/scala/com/nvidia/spark/rapids/shims/spark303/Spark303Shims.scala (hunk 1 above)
Comment (razajafri) [reply sql-plugin/src/main/303/scala/com/nvidia/spark/rapids/shims/spark303/Spark303Shims.scala:2]: Because I am reusing the SparkShimImpl from the folder `301until304`

File: sql-plugin/src/main/31xdb/scala/com/nvidia/spark/rapids/shims/v2/AQEUtils.scala (hunk 6 above)
Comment (razajafri) [reply sql-plugin/src/main/31xdb/scala/com/nvidia/spark/rapids/shims/v2/AQEUtils.scala:28]: Yes, this change is to make the DB build pass. I can spin this out to it's own PR.

File: sql-plugin/src/main/31xdb/scala/com/nvidia/spark/rapids/shims/v2/Spark31XdbShimsBase.scala (hunk 7 above)
Comment (razajafri) [reply sql-plugin/src/main/31xdb/scala/com/nvidia/spark/rapids/shims/v2/Spark31XdbShimsBase.scala:66]: same comment as above

File: sql-plugin/src/main/330/scala/com/nvidia/spark/rapids/v2/RapidsShuffleManager.scala (hunk 10 above)
Comment (razajafri) [reply sql-plugin/src/main/330/scala/com/nvidia/spark/rapids/v2/RapidsShuffleManager.scala:2]: We may be able to. I wasn't sure if I should do it in this PR but sounds like from your earlier comment that you want everything done in one release. I can look into this

**Comment by razajafri** [comment]
build

File: docs/additional-functionality/rapids-shuffle.md [hunk 14]
```diff
@@ -282,30 +282,13 @@ In this section, we are using a docker container built using the sample dockerfi
 
 ### Spark App Configuration
 
-1. Choose the version of the shuffle manager that matches your Spark version.
-   Currently we support:
-
-    | Spark Shim    | spark.shuffle.manager value                              |
-    | --------------| -------------------------------------------------------- |
-    | 3.0.1         | com.nvidia.spark.rapids.spark301.RapidsShuffleManager    |
-    | 3.0.2         | com.nvidia.spark.rapids.spark302.RapidsShuffleManager    |
```
Comment (gerashegalov) [inline docs/additional-functionality/rapids-shuffle.md:291]: How did we end up solving incompatible ShuffleManager base across Apache Spark versions. I thought having a single facade class across shims required dynamic bytecode generation  at Runtime depending on the shim because compile-time method signatures cannot be consistent for each Spark version?

Comment (razajafri) [reply docs/additional-functionality/rapids-shuffle.md]: I tested everything after 3.1.1+ and it worked. I have reverted the change as it wasn't my original intention to remove the duplicate ShuffleManagers. The original intention was to get rid of the dynamic lookup of the shims at runtime.

**Comment by razajafri** [comment]
build

**Comment by razajafri** [comment]
build

File: sql-plugin/src/main/31xdb/scala/com/nvidia/spark/rapids/shims/v2/AQEUtils.scala (hunk 6 above)
Comment (tgravescs) [reply sql-plugin/src/main/31xdb/scala/com/nvidia/spark/rapids/shims/v2/AQEUtils.scala:28]: the DB build fails without these v2 changes?

File: sql-plugin/src/main/301db/scala/com/nvidia/spark/rapids/shims/spark301db/SparkShimServiceProvider.scala (hunk 13 above)
[the remark this answers is not in the document]
Comment (tgravescs) [reply sql-plugin/src/main/301db/scala/com/nvidia/spark/rapids/shims/spark301db/SparkShimServiceProvider.scala:17]: that file needs to be updated for removal of ShimLoader.getSparkShims

File: sql-plugin/src/main/31xdb/scala/com/nvidia/spark/rapids/shims/v2/AQEUtils.scala (hunk 6 above)
Comment (razajafri) [reply sql-plugin/src/main/31xdb/scala/com/nvidia/spark/rapids/shims/v2/AQEUtils.scala:28]: Yes

Comment (tgravescs) [reply sql-plugin/src/main/31xdb/scala/com/nvidia/spark/rapids/shims/v2/AQEUtils.scala:28]: then yes it should be a separate PR and should have details why.. there isn't an open issue for this?  I haven't noticed the DB builds failing but builds have been messed up lately.

**Comment by sameerz** [comment]
build

File: dist/scripts/binary-dedupe.sh [hunk 15]
```diff
@@ -1,6 +1,6 @@
 #!/bin/bash
 
-# Copyright (c) 2021, NVIDIA CORPORATION.
+# Copyright (c) 2021-2022, NVIDIA CORPORATION.
```
Comment (tgravescs) [inline dist/scripts/binary-dedupe.sh:3]: I don't see any other changes to file so remove copy right change

File: sql-plugin/src/main/301/scala/com/nvidia/spark/rapids/spark301/RapidsShuffleManager.scala [hunk 16]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2020-2021, NVIDIA CORPORATION.
+ * Copyright (c) 2020-2022, NVIDIA CORPORATION.
```
Comment (tgravescs) [inline sql-plugin/src/main/301/scala/com/nvidia/spark/rapids/spark301/RapidsShuffleManager.scala:2]: remove, no changes to file

File: sql-plugin/src/main/301db/scala/com/nvidia/spark/rapids/shims/spark301db/RapidsShuffleInternalManager.scala [hunk 17]
```diff
@@ -1,78 +0,0 @@
-/*
- * Copyright (c) 2021, NVIDIA CORPORATION.
```
Comment (tgravescs) [inline sql-plugin/src/main/301db/scala/com/nvidia/spark/rapids/shims/spark301db/RapidsShuffleInternalManager.scala:2]: why was this removed, I don't see one that replaces iit but maybe I'm missing it?

File: sql-plugin/src/main/312-nondb/scala/com/nvidia/spark/rapids/shims/spark312/SparkShimServiceProvider.scala (hunk 5 above)
Comment (tgravescs) [reply sql-plugin/src/main/312-nondb/scala/com/nvidia/spark/rapids/shims/spark312/SparkShimServiceProvider.scala:2]: still needs updated

File: sql-plugin/src/main/320/scala/com/nvidia/spark/rapids/shims/SparkShims.scala [hunk 18]
```diff
@@ -51,7 +50,7 @@ class Spark320Shims extends Spark320PlusShims with Spark30Xuntil33XShims with Re
       lookupFileMeta: String => String,
       dateTimeRebaseModeFromConf: String): ParquetFilters = {
     val datetimeRebaseMode = DataSourceUtils
-      .datetimeRebaseMode(lookupFileMeta, dateTimeRebaseModeFromConf)
+        .datetimeRebaseMode(lookupFileMeta, dateTimeRebaseModeFromConf)
```
Comment (tgravescs) [inline sql-plugin/src/main/320/scala/com/nvidia/spark/rapids/shims/SparkShims.scala:53]: look like to many spaces?

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/ShimLoader.scala [hunk 19]
```diff
@@ -53,13 +53,13 @@ import org.apache.spark.util.MutableURLClassLoader
...
+    jar:file:/home/spark/rapids-4-spark_2.12-22.02.0.jar!/spark3xx-common/
```
Comment (tgravescs) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/ShimLoader.scala:56]: this should have 22.04, revert

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/ShimLoader.scala [hunk 20]
```diff
@@ -99,22 +100,6 @@ object ShimLoader extends Logging {
...
   }
 
-  // Ideally we would like to expose a simple Boolean config instead of having to document
-  // per-shim ShuffleManager implementations:
-  // https://github.com/NVIDIA/spark-rapids/blob/branch-21.08/docs/additional-functionality/
-  // rapids-shuffle.md#spark-app-configuration
-  //
-  // This is not possible at the current stage of the shim layer rewrite because of the combination
-  // of the following two reasons:
-  // 1) Spark processes ShuffleManager config before any of the plugin code initialized
```
Comment (tgravescs) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/ShimLoader.scala:109]: this comment needs to be put back

**Comment by tgravescs** [comment]
build

**Comment by razajafri** [comment]
build

**Comment by razajafri** [comment]
build

**Comment by razajafri** [comment]
build

**Comment by razajafri** [comment]
build

**Comment by razajafri** [comment]
build

**Review by tgravescs — APPROVED** [review]

**Comment by pxLi** [comment]
we do exactly match for post-fix trigger `[databricks]` instead of [Databricks] for db build as https://github.com/NVIDIA/spark-rapids/blob/branch-22.04/CONTRIBUTING.md#blossom-ci

created ticket here https://github.com/NVIDIA/spark-rapids/issues/4918
